spark3/src/main/java/org/apache/iceberg/actions/Spark3CreateAction.java

```diff
@@ -0,0 +1,172 @@
+      this.sourceCatalogTable = sourceTable.v1Table();
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException(String.format("Cannot not find source table %s", sourceTableIdent), e);
+    } catch (ClassCastException e) {
+      throw new IllegalArgumentException(String.format("Cannot use non-v1 table %s as a source", sourceTableIdent), e);
+    }
+    validateSourceTable(sourceCatalogTable);
+
+    this.sourceTableLocation = CatalogUtils.URIToString(sourceCatalogTable.storage().locationUri().get());
+  }
+
+  @Override
+  public CreateAction withProperties(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public CreateAction withProperty(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  protected SparkSession spark() {
+    return spark;
+  }
+
+  protected String sourceTableLocation() {
+    return sourceTableLocation;
+  }
+
+  protected CatalogTable v1SourceTable() {
+    return sourceCatalogTable;
+  }
+
+  protected CatalogPlugin sourceCatalog() {
+    return sourceCatalog;
+  }
+
+  protected Identifier sourceTableName() {
+    return sourceTableIdent;
+  }
+
+  protected Transform[] sourcePartitionSpec() {
+    return sourceTable.partitioning();
+  }
+
+  protected StagingTableCatalog destCatalog() {
+    return destCatalog;
+  }
+
+  protected Identifier destTableName() {
+    return destTableIdent;
+  }
+
+  protected Map<String, String> additionalProperties() {
+    return additionalProperties;
+  }
+
+  private static void validateSourceTable(CatalogTable sourceTable) {
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
+
+    if (!ALLOWED_SOURCES.contains(sourceTableProvider)) {
+      throw new IllegalArgumentException(
+          String.format("Cannot create an Iceberg table from source provider: %s", sourceTableProvider));
+    }
+
+    if (sourceTable.storage().locationUri().isEmpty()) {
+      throw new IllegalArgumentException("Cannot create an Iceberg table from a source without an explicit location");
+    }
+  }
+
+  protected static Map<String, String> tableLocationProperties(String tableLocation) {
+    return ImmutableMap.of(
+        TableProperties.WRITE_METADATA_LOCATION, tableLocation + "/" + ICEBERG_METADATA_FOLDER,
+        TableProperties.WRITE_NEW_DATA_LOCATION, tableLocation
+    );
+  }
+
+  protected static void assignDefaultTableNameMapping(Table table) {
+    NameMapping nameMapping = MappingUtil.create(table.schema());
+    String nameMappingJson = NameMappingParser.toJson(nameMapping);
+    table.updateProperties().set(TableProperties.DEFAULT_NAME_MAPPING, nameMappingJson).commit();
+  }
+
+  private static StagingTableCatalog checkDestinationCatalog(CatalogPlugin catalog) {
+    if (!(catalog instanceof SparkSessionCatalog) && !(catalog instanceof SparkCatalog)) {
```

Comment:
It looks like having a separate method to check if it is an Iceberg catalog would make it more readable than having two negations.

Or we could use `ValidationException`. It also gives formatting for free.

```
    ValidationException.check(
        catalog instanceof SparkSessionCatalog || catalog instanceof SparkCatalog,
        "Unsupported destination catalog. Catalog %s is %s but must be either %s or %s",
        catalog.name(), catalog.getClass().getName(), SparkSessionCatalog.class.getName(), SparkCatalog.class.getName());
```

Reply:
Using `ValidationException` seems pretty clean. What do you think, @RussellSpitzer?

Reply:
Or shall we even use `Preconditions.checkArgument`?